(feat): SEO: Add frontpage meta keywords tags based on all articles

iranzo opened the pull request:
Signed-off-by: Pablo Iranzo Gómez <Pablo.Iranzo@gmail.com>

<!--
    ----------^ Click "Preview" for a nicer view!
-->

<!--
    Thank you very much for contributing to Pelican-Elegant! ❤️
-->

## Prerequisites

- [x] My Pull Request is filled against the `next` branch
- [x] All commits in my Pull Request conform to [Elegant Git commit Guidelines](https://elegant.oncrashreboot.com/git-commit-guidelines)

## Recommended Steps

<!---
    These are not mandatory and will NOT negatively effect our patch review process.
    But we encourage you to do them.
-->

- [ ] My patch adds a new feature, therefore I have also added a help article about it
- [ ] My patch changes Elegant behavior, therefore I have updated the help article to reflect this change
- [x] My commits are [signed](https://help.github.com/en/articles/signing-commits)

## Description

<!--- Provide a general summary of the patch here -->
This adds meta tags for content and keywords plus the default icon for main page of website

Closes #505

iranzo commented:
@talha131 can you please doble check if the 'coding' way would be ok before I take over documenting the new var 'SITE_DESCRIPTION' ?

Thanks,
Pablo

iranzo commented:
One of the issues, btw is that I want this to affect (for meta description) only for 'base', but not for tags, keywords, etc. Any hint?

iranzo commented:
Ok, as SITE_DESCRIPTION is already on NEXT no need to document afaik, it just fills in the extra headers and that's all.

talha131 commented on THANKS.md:
```diff
@@ -30,6 +30,7 @@ Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/d
 
 <!-- markdownlint-enable -->
 <!-- prettier-ignore-end -->
+
```
@iranzo can you please remove this redundant patch?

iranzo replied:
I had to add it to avoid prettier failing the build on travis, I can remove it for sure

iranzo replied:
But when trying to commit, local prettier will change it again

talha131 replied:
I see. Thanks

talha131 commented on templates/base.html:
```diff
@@ -21,9 +21,17 @@
         <meta name="author" content="{{ AUTHOR }}" />
         {% endif %}
 
+        <!-- Fill in Site ICON-->
```
When you put in the base, it will conflict with the feature image of the article.

It should be placed in index.html unless I am missing something.

iranzo replied:
Ok, let me do it

iranzo replied:
In this case, to make it cleaner I would switch to 'base.html' the code so that article uses local one if artciel and use global one if not

iranzo replied:
Moved from the macro that only shows on articles in smo_metadata  to base.html having the if there

talha131 commented on templates/index.html:
```diff
@@ -15,13 +15,10 @@
-{% endif %}
 <meta property="og:site_name" content="{{ SITENAME|striptags|e }}" />
 <meta property="og:article:author" content="{{ AUTHOR }}" />
+<!-- Fill in Site Keywords-->
```
Do we have keywords for a site? To my understanding, keywords are for an article or page but not a whole site. Can you please confirm?

iranzo replied:
My understanding is that it's for the keywords we can use to search for the site, so the 'main page' should be containing all the keywords from all the articles.

talha131 commented on templates/index.html:
```diff
@@ -15,18 +15,11 @@
-<meta name="twitter:image" content="{{FEATURED_IMAGE}}" >
-{% endif %}
+<!-- Fill in Site Keywords-->
+<meta name="keywords" content="{% for tag, articles in tags|sort %} {{ tag }}, {% endfor %}" />
```
If we are not going to use `articles` variable than can we use `_` here so that future maintainers know this value is not used?

iranzo replied:
done in next push

iranzo replied:
Done

talha131 commented on templates/index.html:
```diff
@@ -15,18 +15,11 @@
-<meta name="twitter:image" content="{{FEATURED_IMAGE}}" >
-{% endif %}
+<!-- Fill in Site Keywords-->
+<meta name="keywords" content="{% for tag, articles in tags|sort %} {{ tag }}, {% endfor %}" />
```
Can you please also add categories too in the `keywords` value?

iranzo replied:
I think I did :)

talha131 commented on templates/base.html:
```diff
@@ -21,9 +21,21 @@
         <meta name="author" content="{{ AUTHOR }}" />
         {% endif %}
 
+        {% if article and article.featured_image %}
```
I suggest moving out this patch from base.html to smo_metadata file and then include it in the base.html

iranzo replied:
How should I include it on base so that it works with all elements (article, page, landing etc ? )

iranzo commented:
Also, there are different descriptions for different pages:

archives.html:<meta property="og:description" content="Full archives of {{ SITENAME|striptags|e }} blog" />
archives.html:<meta name="twitter:description" content="Full archives of {{ SITENAME|striptags|e }} blog">
base.html:        <meta name="description" content="{% block head_description %}{{ SITE_DESCRIPTION|e }}{% endblock head_description %}" />
base.html:        <meta name="twitter:description" content="{% block head_description %}{{ SITE_DESCRIPTION|e }}{% endblock head_description %}">
categories.html:<meta property="og:description" content="All categories of the {{ SITENAME|striptags|e }} blog" />
categories.html:<meta name="twitter:description" content="All categories of the {{ SITENAME|striptags|e }} blog">
search.html:<meta property="og:description" content="Search results for {{ SITENAME|striptags|e }} blog" />
search.html:<meta name="twitter:description" content="Search results for {{ SITENAME|striptags|e }} blog">
tags.html:<meta property="og:description" content="All tags used in the {{ SITENAME|striptags|e }} blog" />
tags.html:<meta name="twitter:description" content="All tags used in the {{ SITENAME|striptags|e }} blog">

Not sure on what whould be the preferred approach here

iranzo commented:
As I had no clear idea on how to use the  macro with page, I've simplified this to cover the issue: add the tags for meta on frontpage, covered duplicated and added a missing twitter description.

We can cover the use the smo_template on another issue/pr but I'll need assistance there

talha131 commented on templates/index.html:
```diff
@@ -7,26 +7,22 @@
 {{ super() }}
 {% from '_includes/_defaults.html' import LANDING_PAGE_TITLE with context %}
-{% if LANDING_PAGE_TITLE  %}
+{% if LANDING_PAGE_TITLE %}
```
Nice catch

talha131 commented:
I have merged this PR manually.

@iranzo Please review this commit https://github.com/Pelican-Elegant/elegant/commit/28bcb8592cf595761d96904b004df3f425259849

1. Let me know if you have any question. 
1. Your idea of using `head_description2` was not going to work. User would have end up with two `twitter:description`
1. Let me know if my patch does what you were gunning for.

iranzo commented:
Hi @talha131 the 'head_description' was to deal with the block for the twitter one and avoid issue with two blocks with same name
